### PR TITLE
Some fixes for OPeNDAP attributes

### PR DESCRIPTION
**Pull request by noaaroland**
It seems that double quotes in a String attribute value need to be escaped. The THREDDS Data Server escapes double quotes and the OPeNDAP client I tested (netCDF 4 C-library) fails when reading an attribute with a double quote.

Some old-school netCDF files have np.float32 as attribute values so these need to be accounted for when building the OPeNDAP object (other data types too probably).

### CODE REVIEW

**Comment by abkfenris** [comment]
Thanks for the PR @noaaroland ! 

It would be great to add a test with example problematic attributes, though I think we would need to get #20 working so that we can serve a dataset to clients that would have issues.

**Comment by noaaroland** [comment]
Where does this data set dap_client.get("/datasets/air/opendap.dds") come from? Is it part of the xpublish distribution? 

I have small data files which illustrate the problem. How would I reference them from a test?

**Comment by noaaroland** [comment]
I'm not sure what you're telling me regarding the getting #20 working or how I could help, but it occurs to me that the change I made relates to a data which which at a word in double quotes in a global attribute. See the word history in history_of_appended_files global attribute here: https://data.pmel.noaa.gov/aclim/thredds/dodsC/B10K-K20_CORECFS/Level1/1970-1974/B10K-K20_CORECFS_1970-1974_average_temp.nc.das

Perhaps one could test that case by adding such an attribute to the data set you're building. E.g.

```
@pytest.fixture(scope="session")
def dataset():
    from xarray.tutorial import open_dataset
    ds = open_dataset("air_temperature")
    ds.attrs['quattr'] = 'This attributes uses "quotes"'
    return ds
```

The other modifications I made were because attributes were falling through to the else which were not strings. Eventually I went testing for instanceof str directly, and falling through to the else as it was before. I feel like this might not be ideal in the end.

Let me know specifically how you'd like me to help out and I'll try to help improve the plugin.

**Comment by abkfenris** [comment]
Ok, maybe I didn't get my thoughts down well. 

I just made a [PR](https://github.com/noaaroland/xpublish-opendap/pull/1) to your branch with most of the changes from #20 merged in. I went and disabled the changes from #20 that were causing failing tests (what we were trying to figure out in #18 ), but left the bits active that allow us to run a live server in our testing process. 

That allows us to see how real clients interpret things like attributes (and may help with debugging your question here too https://github.com/xpublish-community/xpublish/discussions/246 ). 

With the test server, I created a [new dataset](https://github.com/xpublish-community/xpublish-opendap/blob/d12c661d471e9fa457099a948a43dbedb73bc2ef/tests/server.py#L9-L10) for the server to test how quotes get formatted in attributes, and a [test that checks that](https://github.com/xpublish-community/xpublish-opendap/blob/d12c661d471e9fa457099a948a43dbedb73bc2ef/tests/test_server.py#L35-L40).

Hopefully that makes a bit more sense.

If that sounds good, you can land https://github.com/noaaroland/xpublish-opendap/pull/1 onto your branch. Then I would suggest that you add datasets and tests for each one of the attribute format conversions to make sure they can be appropriately formatted by the server and loaded on the client.

File: tests/server.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+ds = xarray.tutorial.open_dataset("air_temperature")
+
+ds_attrs_quote = xarray.tutorial.open_dataset("air_temperature")
+ds_attrs_quote.attrs["quotes"] = 'This attribute uses "quotes"'
+ds_attrs_quote.attrs["npint"] = np.int16(16)
+ds_attrs_quote.attrs["npintthirtytwo"] = np.int32(32)
+
+rest = xpublish.Rest(
+    {"air": ds, "attrs_quote_types": ds_attrs_quote},
```
Comment (abkfenris) [inline tests/server.py:16]: Can we split casting and escaping attrs into separate datasets and tests?

```suggestion
ds_attrs_cast = xarray.tutorial.open_dataset("air_temperature")
ds_attrs_cast.attrs["npint"] = np.int16(16)
ds_attrs_cast.attrs["npintthirtytwo"] = np.int32(32)

rest = xpublish.Rest(
    {"air": ds, "attrs_quote_types": ds_attrs_quote, "attrs_cast": ds_attrs_cast},
```

Even though they happen in the same functions, they are different enough that they are likely different enough to fail in different ways. 

If we split them into separate datasets and tests, then we can tell which failed in the future when someone like me starts messing with that function tries to clean things up while forgetting why it's ended up the way it is.

**Comment by aufdenkampe** [comment]
@abkfenris, it seems that the two issues just posted by @dblodgett-usgs might be fixed by this PR:
- #57 
- #58 

Note that the endpoint that @dblodgett-usgs was testing is our deployment of https://github.com/xpublish-experiments/Catalog-To-Xpublish developed by @xaviernogueira and that @sjordan29 will maintain.

**Comment by abkfenris** [comment]
> @abkfenris, it seems that the two issues just posted by @dblodgett-usgs might be fixed by this PR:
> 
>     * [json view of dataset encoded as string? #57](https://github.com/xpublish-community/xpublish-opendap/issues/57)
> 
>     * [Issue escaping string attributes with `"` characters? #58](https://github.com/xpublish-community/xpublish-opendap/issues/58)
> 
> 
> Note that the endpoint that @dblodgett-usgs was testing is our deployment of https://github.com/xpublish-experiments/Catalog-To-Xpublish developed by @xaviernogueira and that @sjordan29 will maintain.
> 
> How might we help move this PR forward?

Whoops, I hadn't seen that there had been more progress here.

The first issue I believe is with a Catalog-to-Xpublish specific route, so I'm going to try to nudge that over there, but you're right in that the second one may be solved by this.

@noaaroland have you managed to deal with your attribute issues with this? Sorry I had missed that you had put further work into this.

**Comment by noaaroland** [comment]
I could get the code to run without error based on the changes I made to the way attributes are treated, but it does still not return the correct OPeNDAP data structures for my use case. See: https://github.com/xpublish-community/xpublish/discussions/246

If you can whip this pull request into shape and use it please do, but I pretty much moved on when I didn't get any feedback.

**Comment by abkfenris** [comment]
Sorry, I don't know quite enough about how Xarray decides if something should be an coordinate and/or dimension when reading OpenDAP, so I pinged some folks offline about https://github.com/xpublish-community/xpublish/discussions/246 I'll try to give folks a nudge again to take a look.

I can try to take care of the PR, but it may take me a few weeks as I'm about to head out on an expedition.

**Comment by noaaroland** [comment]
Thank you. In my understanding of an old-school OPeNDAP server it's not really making an decisions about what is a coordinate, data variable or whatnot. The client figures out what's what by applying the convention by interpreting the attributes and looking at the structure (like is a variable the same name as the dimension). The server is just serving the attributes, structure and hunks of data as the client requests.

In this implementation since it's xarray on both ends, it gets confusing as to which piece is responsible for organizing the xarray object according to the conventions.

I hope we can get help sorting it out. Good luck with your expedition.

**Comment by abkfenris** [comment]
Thank you again @noaaroland for your work on this, and I'll try to get someone to take a look at the dimensions/coordinates issues you've had.

**Pull request merged**